mobify/platform-scaffold
# Web 1284  web 1285  web 1286  fonts

**JIRA**: [WEB-1284](https://mobify.atlassian.net/browse/WEB-1284), [~~WEB-1285~~](https://mobify.atlassian.net/browse/WEB-1285), [~~WEB-1286~~](https://mobify.atlassian.net/browse/WEB-1286)

## Changes
- Add `webfontloader` library
- Defer font loading to a bit later in the app life cycle (when App container mounts)
- Fix FOUT by displaying system fonts until Oswald is loaded, THEN show Oswald fonts, as per [Zack Leatherman's method here](https://www.zachleat.com/web/comprehensive-webfonts/#fout-class)

## Impact

### Animated!

Notice how in _After_ the `Shop` text is always visible. In both cases, the category name is placeholder if the text is not available.

| Before | After |
| ------ | ------ |
| ![fout-before](https://cloud.githubusercontent.com/assets/734535/23976726/e8f2bde6-09a6-11e7-9b82-ed09d731ad1c.gif) | ![fout-after](https://cloud.githubusercontent.com/assets/734535/23976728/ed28b370-09a6-11e7-95b4-a797046a9061.gif) |

### First Paint of Text

Notice how there text in the _Before_ is invisible. At the same point in time, the _After_ screenshot has text perfectly visible!

| Before | After |
| ------ | ------ |
| ![earliest-text-before](https://cloud.githubusercontent.com/assets/734535/23976482/3f262c7c-09a5-11e7-9ae7-3c495c448b88.png) | ![earliest-text-after](https://cloud.githubusercontent.com/assets/734535/23976488/43b04bd8-09a5-11e7-9433-1059e1582317.png) |

### Network

Notice how the loading of fonts has been deferred to a later point in time. In the _Before_ screen, you can see that the Google API CSS file starts to downloaded after `main.css` finishes downloaded (that's where we were importing the fonts originally) and the font files themselves download before content is rendered (we can assume this is the case because image assets haven't started downloading yet).

In the _After_, the Google API CSS downloads at the same time as we fetch the SVG asset. The font files themselves end up downloading just after image assets.

I assert that the overall impact this will have is that large impact assets will load first, and lesser impact assets (fonts) will load slightly later. I hope this means visual progression as measured by Web Page Test (i.e. page speed index) will improve by this change.

| Before | After |
| ------ | ------ |
| ![network-before](https://cloud.githubusercontent.com/assets/734535/23976491/4908feea-09a5-11e7-98cf-e09c60e22d14.png) | ![network-after](https://cloud.githubusercontent.com/assets/734535/23976494/4ce1095e-09a5-11e7-9bb8-4a10603aa7e5.png) |

## How to test-drive this PR
- Run `npm run dev`
- Preview [Merlin's Potions](https://preview.mobify.com/?url=https%3A%2F%2Fwww.merlinspotions.com%2F&site_folder=https%3A%2F%2Flocalhost%3A8443%2Floader.js&disabled=0&domain=&scope=1)
- Verify that fonts load as intended
